training/eval/bucketing.py

```diff
@@ -0,0 +1,231 @@
+    Maps model value to int label by provided boundaries.
+    Args:
+        x: value which has to be mapped
+        boundaries: any long list of floats which describes boundaries of
+            mapping buckets e.g. [0.33, 0.54] where
+            label 1 <= 0.33 < label 2 <= 0.54 < label 3
+    """
+    for idx, bound in enumerate(boundaries):
+        if x <= bound:
+            return idx + 1
+    return len(boundaries) + 1
+
+
+def optimize_buckets(
+    df: pd.DataFrame,
+    optimization_column: str,
+    human_prefix: str = "label_",
+    n_labels: int = 5,
+    verbose: bool = False,
+) -> Tuple[List[float], float]:
+    """
+    Oprimizes mapping buckets for n labels for given models continuous outputs.
+    Args:
+        df: DataFrame where each row is a one example and each column is
+            an annotator. The values of human labels should not be normalized.
+        optimization_column: column with values under which we want to
+            optimize labels mapping (values should be normalized to range 0-1)
+        human_prefix: prefix of the columns' names that store the reference
+            annotations (human labels).
+        n_labels: number of labels to which we want to map the model's
+            responses.
+        verbose: whether to print optimization progress.
+    Returns:
+        boundaries: List of bounds for the buckets.
+        max_score: The score of the best mapping.
+    """
+    df = df.copy()
+    max_score = -np.inf
+    previous_bound = 0
+    boundaries = []
+    for bound_num in range(n_labels - 1):
+        if verbose:
+            print(f"Optimizing {bound_num + 1} boundary")
+        temp_boundaries = boundaries.copy()
+        temp_boundaries.append(previous_bound)
+        for bound in np.arange(previous_bound, 1, 0.01):
+            temp_boundaries[bound_num] = bound
+            df["model"] = df[optimization_column].apply(
+                lambda x: map_to_optimal_bucket(x, temp_boundaries)
+            )
+            score = _get_labeler_score_optimization_version(
+                df, "model", human_prefix, len(temp_boundaries) + 1
+            )
+            if bound_num == 0:
+                if score >= max_score:
+                    if len(boundaries) == bound_num:
+                        boundaries.append(bound)
+                    else:
+                        boundaries[bound_num] = bound
+                    max_score = score
+            else:
+                if score > max_score:
+                    if len(boundaries) == bound_num:
+                        boundaries.append(bound)
+                    else:
+                        boundaries[bound_num] = bound
+                    max_score = score
+        max_score = -np.inf
+        previous_bound = boundaries[bound_num] + 0.01
+    return boundaries, max_score
+
+
+def _get_labeler_score_optimization_version(
+    df: pd.DataFrame,
+    labeler_col: str,
+    human_prefix: str = "label_",
+    n_labels: int = 5,
+):
```

Comment:
nitpick: missing return type hint, it's not required here, but on the gitlab I think there are precommit hooks that would fail in this case

Reply:
there are types missing in `_get_tr_df_wide` and `_to_wide` as well